### ISSUE TITLE
CoordSystem's scalars needs assumptions

### ISSUE THREAD

**Issue by mcpl-sympy** [sympy/sympy#19451]
`CoordSystem` in `diffgeom` module has `_dummies` attribute, which represents its base scalars. However, these symbols does not have any assumptions.

```python
>>> from sympy.diffgeom.rn import R2_r, R2_p
>>> R2_r._dummies[0].is_real    # Cartesian's x should be real, but None returned
>>> R2_p._dummies[0].is_nonnegative    # Cartesian's r should be nonnegative, but None returned
```

I have an idea of solving this:  
First, Instead of `Dummy`, introduce `BaseScalar` class (subclass of `Symbol`) and use these as scalar symbols. (Its class name is tentative)  
Second, make `Symbol`s with assumptions passed to `CoordSystem`, which are converted to `BaseScalar` and stored.

```python
>>> from sympy.diffgeom import *
>>> from sympy.diffgeom import R2_origin
>>> x,y = symbols('x y', real=True)
>>> r, theta = symbols('x theta', nonnegative=True)
>>> Car = CoordSystem('Car', R2_origin, [x, y])
>>> Pol = CoordSystem('Pol', R2_origin, [r, theta])

>>> Car.symbols
(x, y)
>>> Car.symbols[0].func
ScalarSymbol
>>> Car.symbols[0].is_real
True

>>> Pol.symbols
(r, theta)
>>> Pol.symbols[0].is_nonnegative
True
```

**Comment by mcpl-sympy** [comment]
Ping @sylee957

**Comment by sylee957** [comment]
Does non-negativeness follows from stuff like coordinate transformations?
If this is the case, there can also be some computational ways to figure out if the assumptions provided to the symbols in CoordSystem are reasonable or not.

**Comment by mcpl-sympy** [comment]
@sylee957 
Yes. "Polar's `r` is nonnegative" comes from the fact that "`r = sqrt(x**2+y**2)`, and `x`,`y` are real".

I tried the approach you suggested, but with current SymPy's assumption system there are only so many that can be deduced.  
```python
>>> x,y = symbols('x y', real=True)
>>> sqrt(x**2 + y**2).is_nonnegative # Polar's r
True
>>> atan2(y,x) # Polar's theta : returns None
```

**Comment by sylee957** [comment]
On the other hand, will there any way to inherit the assumptions of the given coordinate symbols of `CoordSystem` in `BaseScalarField`. 
I guess `CoordSystem.names` can be wrapped with user provided symbols than strings, so this may need to be considered unless introducing any new class is necessary for this.

**Comment by mcpl-sympy** [comment]
@sylee957 
A fair point. We can pass symbols with assumptions to `names` and use these as coordinate symbols.

**Comment by mcpl-sympy** [comment]
Or, since this assumptions are dependent to the context of how systems are related, we can pass the symbols to `CoordSystemStructure` suggested in #19368

**Comment by sylee957** [comment]
From `CoordSystemStructure`, the only concern I have is that you have to specify all 9 assumptions for 3D systems in one place, and I think that it is difficult to come up with an idea to define the signature elegantly.

**Comment by sylee957** [comment]
For `CoordSystemStructure`, you may think of a third matrix or dictionary to store the assumptions, because sympy objects need to store every information explicitly in class signature, and the other concerns can arise, like what if I have to define assumptions partially, or such.

But at least from `CoordSystem`, it always had to function as a container store the symbols for names, so it is not much complicated to carry on to contain names with assumptions.

Although the coupling between CoordSystem and CoordSystemStructure may be weakened after starting to store some informations about the overall systems independently,
I don't have worries of something getting broken after assumptions are changed, because since both CoordSystem and CoordSystemStructure are immutable, you cannot change the assumptions in-place for already created CoordSystem, and already created CoordSystemStructure will always reference the one that were created more older.

**Comment by sylee957** [comment]
The other thing I have worried about is, if someone wants to implement a solver for 
> "Polar's r is nonnegative" comes from the fact that "r = sqrt(x**2+y**2), and x,y are real".

that figures out the assumptions for coordinate systems

But I may also expect that to be done with `fill_in_assumptions(CoordSysStructure, CoordSys1, CoordSys2, ...)` and return a new tuple `NewCoordSysStructure, NewCoordSys1, NewCoordSys2` with filled-out assumptions, so that may not become a new problem.

**Comment by mcpl-sympy** [comment]
@sylee957 
I agree on that we should make the class signature as simple as possible. But I think that assumptions can be passed to `CoordSystemStructure` without complicating the signature significantly - insomuch that we should disturb the coupling between `CoordSystem` and `CoordSystemStructure`.

I'll first proceed with the original plan and show you the result. We can discuss and change it afterhand.

**Comment by mcpl-sympy** [comment]
Also, I think the idea suggested in issue #18028 that you opened is crucial to implement more robust classes in this module. I am planing to take on the job as soon as possible.

**Comment by sylee957** [comment]
Okay, I thought that it was okay to merge #19368 in step and proceed with assumptions, but if you are willing to showcase how the new assumptions works, it can get to more heavier PR.

I was only waiting for your opinion about https://github.com/sympy/sympy/pull/19368#issuecomment-637215547

**Comment by mcpl-sympy** [comment]
@sylee957 I decided that since we're introducing a new design and writing the documents, it would be better to include major changes like this into one PR.  
Unfortunately, my current circumstance impedes me from dedicating into contribution. I'll push and ping you later. Thanks.
